### ISSUE TITLE
Disable idempotence check on certain tasks

### ISSUE THREAD

**Issue by Deepakkothandan** [ansible/molecule#816]
# Issue Type

- Question
- Feature request

# Molecule and Ansible details

```
ansible --version
2.3.0
molecule --version
1.23.0
```

- Molecule installation method: pip
- Ansible installation method:  pip

# Desired Behaviour
Molecule idempotence fails when play contains shell module with non-idempotent behaviour, which is understandable. But, In my use case, I am running a grunt build task, using the shell module, which is non-idempotent in nature.

Is there already a way to exclude tasks from molecule's idempotence check? If not, this would be a nice feature to have. For now, I am running molecule with idempotence disabled.

# Actual Behaviour (Bug report only)

provide a method to exclude tasks from idempotence check, which are not idempotent, due to requirements.

**Comment by Deepakkothandan** [comment]
A task can contains **change_when: False**, which would always report the task to be unchanged on playbook execution. But, one would want to see changed tasks output during execution.

**Comment by retr0h** [comment]
@Deepakkothandan Yeah, this isn't something we would support.  Your automation and tasks should be written in such a way that it is idempotent.  This may require more work on your part by statting certain files or directories before running your grunt build.  Either way, it is up to the user to write automation in an idempotent way.

This is the opinionated portion of Molecule 😄

**Comment by ssbarnea** [comment]
Hmm, I think we may need to reopen this and at least allow one option to disable idempotency check. I think it would be impossible for some roles to be idempotent anyway. The `changed_when: False` hack is clearly not acceptable because it would be clear lie and would have other undersired side effects (not to mention that people are using the changed status to make other decisions later).

**Comment by madrover** [comment]
The PR #1663 should be fixing this.

**Comment by baccenfutter** [comment]
@retr0h seems there may be a valid use-case, after all?

Example:

Picture a role with a variable `users`, which contains a list of users to provision via Ansible's `user` module. The same role also has a variable `users_absent` which defines the list of users to de-provision. When I write tests for my role, I first provision a test user to be able to de-provision it in the same run and then verify that the user does indeed not exist in the test box. If there was an error in my role, the user would be provision and still present by the time `verify` runs.

However, `molecule` never reaches the `verify` stage, because `idempotence` fails.

How would you suggest to approach this without some way of telling Molecule that the two tasks are indeed purposely not idem-potent?

**Comment by clrosier** [comment]
Also would like to pitch the example:

In the case where we use Ansible as a provisioner within something like Packer's docker builder, usually there's a step where you provision Python using the 'raw' module within the docker image so that Ansible can do it's thing and then you clean up Python afterwards to save on space.

If we're using Molecule alongside this method, idempotence will always fail since python is cleaned up at the end and we won't be able to use a 'when' statement to check if python is already installed.

Maybe someone could advise an alternative method here?

**Comment by ssbarnea** [comment]
If you really hate idempotence testing so much you can always override the test_sequence and remove the idempotence from it. See https://molecule.readthedocs.io/en/stable/configuration.html#scenario

**Comment by raizyr** [comment]
> If you really hate idempotence testing so much you can always override the test_sequence and remove the idempotence from it. See https://molecule.readthedocs.io/en/stable/configuration.html#scenario

The problem is that this turns off the test for everything, when what people are asking for is to be able to ignore the idempotence for a single task.  In some cases it might be necessary to capture the output of a query command in order to control idempotence on a later task.  The command output capturing task is always going to run and change and that's okay as it's just registering a variable for the play. 

If you use the `molecule-idempotence-notest` then the task is skipped causing later plays to fail because the variable isn't registered.

```
- name: Get Swap Status # noqa 301
  shell: |
    set -o pipefail
    swapon --show --noheadings | awk '{print $1}'
  args:
    executable: /bin/bash
  register: swapon_show_output
  tags:
    - molecule-idempotence-notest

- name: Set swapfiles fact
  set_fact:
    swapfiles: "{{ swapon_show_output.stdout_lines | list }}"
```

Also, unfortuantely `|` in set_fact doesn't work and that task still fails because the variable is unregistered.

**Comment by pellywog** [comment]
why is this issue closed?

**Comment by ssbarnea** [comment]
That is because you can use changed_when to change the behaviour in Ansible, or you can completely remove idenpotence step from the test sequence.

Basically there are already two ways to address it.

**Comment by cdaniluk** [comment]
It's worth pointing out that neither addresses the actual issue. Some specific tasks will inherently fail an idempotence test for valid reasons. That doesn't mean that setting `changed_when: False` always makes sense either, though. Sometimes an idempotence test can fail yet `changed_when` should not be set to false. Saying that you can disable _all_ idempotence testing for a playbook to deal with it isn't really an answer either. Ideally, Molecule would have some way (perhaps via config) to have a list of tasks to exclude idempotence checks from.

**Comment by madrover** [comment]
What I use is:
```
changed_when: "'molecule-idempotence-notest' not in ansible_skip_tags"
```
This executes the task during the *idempotence* phase, thus catching possible errors, and, at the same time, preventing the phase to fail due unwanted changes.

**Comment by pellywog** [comment]
Wasn't quite working for me, so I tried to find out why:

`ansible_skip_tags` has some defaults and can only be overridden by command line flags, not by specifying it in the playbook or via set_fact or similar: https://serverfault.com/questions/936611/overriding-ansible-skip-tags-within-a-
playbook

Here is an example of what's in it:
```
# play 
- name: print ansible_skip_tags
  debug:
    msg: "{{ ansible_skip_tags }}"

# result
TASK [role-ssh : print ansible_skip_tags] ***********************************************************
ok: [bullseye] =>
  msg:
    - molecule-notest
    - notest
```

So I am using this to "detect if the play is running in molecule or not":
```
- name: do something that is not idempotent
  shell: echo foobar
  changed_when: "'molecule-notest' not in ansible_skip_tags"
```

This marks the task as not changed when it is running in molecule.            

What would be the correct way to add "molecule-idempotency-notest" to the list "ansible_skip_tags"? I assume via molecule.yml somehow?

**Comment by pellywog** [comment]
I'd still like to point out that `changed_when` is not a pretty solution for the problem.. It should be possible to just exclude tasks from the idempotency check. `changed_when` will just mark the task as not executed in Ansible if its running in molecule, which is not really the desired behavior.

I think this issue should be re-opened @ssbarnea

**Comment by jaybaws** [comment]
I agree with @pellywog. Why can't we add some sort of hint in the task name, to let Molecule know it should ignore idempotence issues for that particular task? A similar feature as ansible-lint's `noqa` comments would be appreciated.

**Comment by maximmasiutin** [comment]
By the way, there is a very good article on idempotence of Ansible playbooks at https://datamattsson.tumblr.com/post/186523898221/idempotence-and-changedwhen-with-ansible

And there are some minor things on idempotence explained at:
https://univers-libre.net/posts/ansible-molecule.html
https://molecule.readthedocs.io/en/latest/faq.html

**Comment by gwvandesteeg** [comment]
So the question I have regarding this is for example the `ufw` module, let's say you want to ensure that you start with a clean slate of firewall rules and as such use.

```
- name: Reset firewall rules to ensure we start with a blank slate
  ufw:
    state: reset
```
and then load a bunch of firewall rules.  The above statement is not idempotent, however without it you cannot guarantee that the rules you are applying are the only rules on the system in question.  But by applying this task during the idempotence check you are removing all applied firewall rules and as such all firewall rules will come back again as `changed`.

**Comment by xcorp** [comment]
Another good example is restarting a service with systemd to make sure config changes is applied, this will always returned changed.

**Comment by mafalb** [comment]
Yes, but there is the molecule-idempotence-notest tag which was implemented in #1663

**Comment by xcorp** [comment]
Yeah, I saw that later :)
